src/main/java/com/google/devtools/build/lib/query2/cquery/GraphOutputFormatterCallback.java

```diff
@@ -0,0 +1,115 @@
+// Copyright 2020 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.query2.cquery;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.devtools.build.lib.analysis.ConfiguredTarget;
+import com.google.devtools.build.lib.cmdline.Label;
+import com.google.devtools.build.lib.events.ExtendedEventHandler;
+import com.google.devtools.build.lib.graph.Digraph;
+import com.google.devtools.build.lib.graph.Node;
+import com.google.devtools.build.lib.query2.engine.QueryEnvironment.TargetAccessor;
+import com.google.devtools.build.lib.query2.query.output.GraphOutputWriter;
+import com.google.devtools.build.lib.query2.query.output.GraphOutputWriter.NodeReader;
+import com.google.devtools.build.lib.skyframe.SkyframeExecutor;
+import java.io.OutputStream;
+import java.util.Comparator;
+
+/** cquery output formatter that prints the result as factored graph in AT&amp;T GraphViz format. */
+class GraphOutputFormatterCallback extends CqueryThreadsafeCallback {
+  @Override
+  public String getName() {
+    return "graph";
+  }
+
+  /** Interface for finding a configured target's direct dependencies. */
+  @FunctionalInterface
+  public interface DepsRetriever {
+    Iterable<ConfiguredTarget> getDirectDeps(ConfiguredTarget taget) throws InterruptedException;
+  }
+
+  private final DepsRetriever depsRetriever;
+
+  private final GraphOutputWriter.NodeReader<ConfiguredTarget> nodeReader =
+      new NodeReader<ConfiguredTarget>() {
+
+        private final Comparator<ConfiguredTarget> configuredTargetOrdering =
```

Comment:
For my own edification - reason to initialize this outside of the comparator method?

Reply:
My original reasoning is that the `comparator` method can be called multiple times whereas the actual comparator logic only needs to be defined once. So it theoretically saves unnecessary extra instantiation.

In practice I don't think that'd make a huge difference (and I'd hope the JDK could optimize that). So I don't have strong feelings on the subject.